### PR TITLE
`Integrated code lifecycle`: Decouple submission result processing from build agent logic

### PR DESCRIPTION
**Pull request by laurenzfb**
# Do NOT deploy on testservers, contains config change

#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Description

This PR further decouples the build agent logic from the general Artemis code:
- LocalCIBuildResults and other information are now pushed into a distributed queue by the build agent
- Artemis nodes pull items from this queue, grade the submission and notify the user
- Build scripts are no longer temporarely stored as files and directly send to the build agents via the build job queue

Effective with this PR, build agents do not need access to the database anymore. The need for access to the shared file system is also further reduced and will be eliminated in a follow-up PR.

### Steps for Testing
**Only test locally**
The "Artemis (Server, LocalVC & LocalCI)" run configuration should have a new profile "buildagent". Check if it is there, if not add it manually

Prerequisites:
- 1 Instructor

1. Create different kinds of programming exercises and check whether the build results are as expected

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
    - Introduced serialization support for certain data transfer objects to enhance data handling.
    - Added new service for processing build job results, including notifying users and saving results to the database.
    - Implemented functionality for managing and processing build jobs from a shared queue, improving build job handling efficiency.
- **Refactor**
    - Updated various services and classes to use `SharedQueueManagementService` for better build job queue management.
    - Modified the `createBuildScript` method to return a build script string, simplifying the build script generation process.
    - Renamed and updated classes and methods to align with the new "buildagent" profile, enhancing clarity and consistency across the codebase.
- **Chores**
    - Added the "buildagent" profile to configurations and tests, aligning the application setup with the new build agent infrastructure.
- **Bug Fixes**
    - Fixed the handling of build scripts within Docker containers, ensuring proper script creation and execution.

### CODE REVIEW

**Review by georgyia — DISMISSED** [review]
manually tested on ts2, works as expected

**Review by Strohgelaender — COMMENTED** [review]
Had a first look over the server code and added some comments / questions:

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobContainerService.java [hunk 1]
```diff
@@ -82,22 +82,24 @@ public LocalCIContainerService(DockerClient dockerClient, HostConfig hostConfig,
...
                 // Command to run when the container starts. This is the command that will be executed in the container's main process, which runs in the foreground and blocks the
                 // container from exiting until it finishes.
                 // It waits until the script that is running the tests (see below execCreateCmdResponse) is completed, and until the result files are extracted which is indicated
                 // by the creation of a file "stop_container.txt" in the container's root directory.
-                .withCmd("sh", "-c", "while [ ! -f " + LOCALCI_WORKING_DIRECTORY + "/stop_container.txt ]; do sleep 0.5; done")
-                // .withCmd("tail", "-f", "/dev/null") // Activate for debugging purposes instead of the above command to get a running container that you can peek into using
+                // .withCmd("sh", "-c", "while [ ! -f " + LOCALCI_WORKING_DIRECTORY + "/stop_container.txt ]; do sleep 0.5; done")
+                .withCmd("tail", "-f", "/dev/null") // Activate for debugging purposes instead of the above command to get a running container that you can peek into using
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/BuildJobContainerService.java:102]: please change that back

**Review by reschandreas — CHANGES_REQUESTED** [review]
tested locally, works as expected

**Review by rstief — DISMISSED** [review]
tested locally for python, worked as expected.

**Review by milljoniaer — APPROVED** [review]
Tested in testing session

**Review by milljoniaer — APPROVED** [review]
Reapprove after solving merge conflicts

**Review by rstief — APPROVED** [review]
Reapprove after merge

**Review by reschandreas — APPROVED** [review]
code lgtm

**Review by mateusmm01 — APPROVED** [review]
Code lgtm! Also tested locally and everything still works as expected

**Pull request merged**